pgrx/src/datum/array.rs

```diff
@@ -628,6 +606,10 @@ impl<'a, T: FromDatum> Iterator for ArrayIterator<'a, T> {
         let Some(is_null) = array.null_slice.get(*curr) else { return None };
         *curr += 1;
 
+        // Make sure a previous iter didn't push out of bounds!
+        if !is_null {
+            debug_assert!(*ptr < array.raw.end_ptr());
```

Comment:
Note that the comparison is wrong if `ptr` wrapped around the address space and other such edge cases. Might not be worth giving a shit about, since handling that correctly is annoying.

Actually, I have some code in a non-work project that cares about that, so I tweaked it for pgrx (hopefully not breaking it in the process) which gives <https://gist.github.com/thomcc/b2730e3b1a56c4d8cdebf29814ea2351>. It can probably be simplified for our use though (we perhaps don't need the `size` param), and has not been tested after I tweaked it for pgrx.

Reply:
If you go that route, we have a couple other places where we have similar debug assertions comparing against `end_ptr`, might be worth fixing them too.

Reply:
Can we land that in a separate PR? I probably want it but we should discuss it as its own beast because that raises questions about more of our codebase and this patch has already scope-creeped a bit.

Reply:
Sure.